AppController/djinn.rb

```diff
@@ -5418,25 +5317,93 @@ def find_lowest_free_port(starting_port, ending_port=0, appid="")
   end
 
 
-  # Adds or removes AppServers within a node based on the number of requests
-  # that each application has received as well as the number of requests that
-  # are sitting in haproxy's queue, waiting to be served.
-  def scale_appservers_within_nodes
+  # Scale AppServers up/down for each application depending on the current
+  # queued requests and load of the application.
+  #
+  # Returns:
+  #   An Integer indicating the number of AppServers that we couldn't
+  #   start for lack of resources.
+  def scale_appservers
+    needed_appservers = 0
     @apps_loaded.each { |app_name|
       initialize_scaling_info_for_app(app_name)
 
       # Get the desired changes in the number of AppServers.
       delta_appservers = get_scaling_info_for_app(app_name)
       if delta_appservers > 0
         Djinn.log_debug("Considering scaling up app #{app_name}.")
-        try_to_scale_up(app_name, delta_appservers)
+        unless try_to_scale_up(app_name, delta_appservers)
+          needed_appservers += delta_appservers
+        end
       elsif delta_appservers < 0
         Djinn.log_debug("Considering scaling down app #{app_name}.")
         try_to_scale_down(app_name, delta_appservers.abs)
       else
         Djinn.log_debug("Not scaling app #{app_name} up or down right now.")
       end
     }
+
+    return needed_appservers
+  end
+
+
+  # Adds or removes AppServers and/or nodes to the deployment, depending
+  # on the statistics of the application and the loads of the various
+  # services.
+  def scale_deployment
+    needed_appservers = scale_appservers
+    if needed_appservers > 0
+      Djinn.log_debug("Need to start VMs for #{needed_appservers} more AppServers.")
+    end
+
+    # Here we cound the number of machines we need to spawn, and the roles
+    # we need.
+    vms_to_spawn = 0
+    roles_needed = {}
+    if needed_appservers > 0
+      Integer(needed_appservers/3).downto(0) {
+        vms_to_spawn += 1
+        roles_needed["appengine"] = [] unless roles_needed["appengine"]
+        roles_needed["appengine"] << "node-#{vms_to_spawn}"
+      }
+    end
+
+    # Check if we need to spawn VMs and the InfrartructureManager is
+    # available to do so.
+    return unless vms_to_spawn > 0
+    if SCALE_LOCK.locked?
+      Djinn.log_debug("Another thread is already working with the InfrastructureManager.")
+      return
+    end
+
+    Thread.new {
+      SCALE_LOCK.synchronize {
+        Djinn.log_info("We need #{vms_to_spawn} more VMs.")
+
+        if Time.now.to_i - @last_scaling_time < (SCALEUP_THRESHOLD *
+                SCALE_TIME_MULTIPLIER * DUTY_CYCLE)
+          Djinn.log_info("Not scaling up right now, as we recently scaled " +
+            "up or down.")
+          return
+        end
+
+        allowed_vms = Integer(@options['max_images']) - @nodes.length
```

Comment:
Since `@nodes` is not changed in this same sync block, it seems like this still has the potential for exceeding `max_images`. Is there some global state somewhere that keeps track of machines that are still starting?